File: application/core/src/main/java/com/aidriven/core/tenant/TenantRegistry.java
package com.aidriven.core.tenant;

import lombok.extern.slf4j.Slf4j;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Registry for tenant configurations.
 * Supports dynamic registration and lookup of tenant configs.
 *
 * <p>In production, tenants are loaded from DynamoDB or environment config.
 * In tests, tenants can be registered programmatically.</p>
 */
@Slf4j
public class TenantRegistry {

    private final Map<String, TenantConfig> tenants = new ConcurrentHashMap<>();

    /**
     * Registers a tenant configuration.
     *
     * @param config The tenant configuration to register
     * @throws IllegalArgumentException if tenantId is null or blank
     */
    public void register(TenantConfig config) {
        if (config == null || config.getTenantId() == null || config.getTenantId().isBlank()) {
            throw new IllegalArgumentException("TenantConfig must have a non-blank tenantId");
        }
        tenants.put(config.getTenantId(), config);
        log.info("Registered tenant: {} ({})", config.getTenantId(), config.getTenantName());
    }

    /**
     * Looks up a tenant by ID.
     *
     * @param tenantId The tenant identifier
     * @return Optional containing the tenant config, or empty if not found
     */
    public Optional<TenantConfig> getTenant(String tenantId) {
        if (tenantId == null || tenantId.isBlank()) {
            return Optional.empty();
        }
        return Optional.ofNullable(tenants.get(tenantId));
    }

    /**
     * Returns all registered tenants.
     */
    public Collection<TenantConfig> getAllTenants() {
        return Collections.unmodifiableCollection(tenants.values());
    }

    /**
     * Returns the number of registered tenants.
     */
    public int size() {
        return tenants.size();
    }

    /**
     * Removes a tenant from the registry.
     *
     * @param tenantId The tenant identifier to remove
     * @return true if the tenant was removed, false if not found
     */
    public boolean deregister(String tenantId) {
        boolean removed = tenants.remove(tenantId) != null;
        if (removed) {
            log.info("Deregistered tenant: {}", tenantId);
        }
        return removed;
    }

    /**
     * Resolves the tenant for a given Jira project key.
     * Searches all tenants for one that includes the project key.
     *
     * @param jiraProjectKey The Jira project key (e.g., "PROJ")
     * @return Optional containing the matching tenant, or empty if not found
     */
    public Optional<TenantConfig> resolveByJiraProject(String jiraProjectKey) {
        if (jiraProjectKey == null || jiraProjectKey.isBlank()) {
            return Optional.empty();
        }
        return tenants.values().stream()
                .filter(t -> t.getJiraProjectKeys() != null
                        && t.getJiraProjectKeys().contains(jiraProjectKey))
                .findFirst();
    }
}

File: core/tenant/src/main/kotlin/com/workflow/tenant/TenantRegistry.kt
package com.workflow.tenant

import com.workflow.api.TenantId
import org.slf4j.LoggerFactory
import java.util.concurrent.ConcurrentHashMap

/**
 * Central registry for tenant configurations.
 * Supports dynamic registration and lookup of tenant configs.
 */
interface TenantRegistry {
    fun register(config: TenantConfiguration)
    fun get(tenantId: TenantId): TenantConfiguration
    fun findOrNull(tenantId: TenantId): TenantConfiguration?
    fun getAll(): List<TenantConfiguration>
    fun deregister(tenantId: TenantId)
    fun exists(tenantId: TenantId): Boolean
}

/**
 * In-memory implementation of TenantRegistry.
 * Suitable for single-node deployments; can be replaced with distributed cache.
 */
class InMemoryTenantRegistry : TenantRegistry {

    private val log = LoggerFactory.getLogger(InMemoryTenantRegistry::class.java)
    private val tenants = ConcurrentHashMap<String, TenantConfiguration>()

    override fun register(config: TenantConfiguration) {
        tenants[config.tenantId.value] = config
        log.info("Registered tenant: {} (plan={})", config.tenantId.value, config.plan)
    }

    override fun get(tenantId: TenantId): TenantConfiguration =
        findOrNull(tenantId) ?: throw TenantNotFoundException(tenantId)

    override fun findOrNull(tenantId: TenantId): TenantConfiguration? =
        tenants[tenantId.value]

    override fun getAll(): List<TenantConfiguration> = tenants.values.toList()

    override fun deregister(tenantId: TenantId) {
        tenants.remove(tenantId.value)
        log.info("Deregistered tenant: {}", tenantId.value)
    }

    override fun exists(tenantId: TenantId): Boolean = tenants.containsKey(tenantId.value)
}

class TenantNotFoundException(tenantId: TenantId) :
    RuntimeException("Tenant not found: ${tenantId.value}")
